playbook/pg_listeners.go

```diff
@@ -0,0 +1,89 @@
+package playbook
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/duty/models"
+	"github.com/flanksource/incident-commander/api"
+	v1 "github.com/flanksource/incident-commander/api/v1"
+	"github.com/flanksource/incident-commander/db"
+	"github.com/flanksource/incident-commander/utils"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"gorm.io/gorm"
+)
+
+func ListenPlaybookPGNotify(db *gorm.DB, pool *pgxpool.Pool) {
+	var (
+		dbReconnectMaxDuration         = time.Minute
+		dbReconnectBackoffBaseDuration = time.Second
+	)
+
+	pgNotifyPlaybookSpecApprovalUpdated := make(chan string)
+	go utils.ListenToPostgresNotify(pool, "playbook_spec_approval_updated", dbReconnectMaxDuration, dbReconnectBackoffBaseDuration, pgNotifyPlaybookSpecApprovalUpdated)
+
+	pgNotifyPlaybookApprovalsInserted := make(chan string)
+	go utils.ListenToPostgresNotify(pool, "playbook_approval_inserted", dbReconnectMaxDuration, dbReconnectBackoffBaseDuration, pgNotifyPlaybookApprovalsInserted)
+
+	ctx := api.NewContext(db, nil)
+	for {
+		select {
+		case id := <-pgNotifyPlaybookSpecApprovalUpdated:
+			if err := onApprovalUpdated(ctx, id); err != nil {
+				logger.Errorf("%v", err)
+			}
+
+		case id := <-pgNotifyPlaybookApprovalsInserted:
+			if err := onPlaybookRunNewApproval(ctx, id); err != nil {
+				logger.Errorf("%v", err)
+			}
+		}
+	}
+}
+
+// onApprovalUpdated is called when the playbook spec approval is updated
+func onApprovalUpdated(ctx *api.Context, playbookID string) error {
```

Comment:
Is this the only place the approvals are managed? We can't use pg_notify as the only source as we will miss event when incident commander is down

Reply:
Right! I'll add new sync consumers but sync consumers are implemented in a later PR. I'll make the change there.